### ISSUE TITLE
Values list "r" does not have a column named "source_dataset" - New in 3.9.3

### ISSUE THREAD

**Issue by mshearer0** [moj-analytical-services/splink#1414]
### What happens?

When executing 

df_new = linker.find_matches_to_new_records([record], blocking_rules=[]).as_pandas_dataframe()

I get:

Error was: Binder Error: Values list "r" does not have a column named "source_dataset"

Works fine with 3.9.2 but breaks on recently released 3.9.3

### To Reproduce

See final cell in Chapter6.ipynb at https://github.com/mshearer0/HandsOnEntityResolution.

### OS:

Windows

### Splink version:

3.9.3

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by RobinL** [comment]
Thanks - I've figured out the source of this so we will fix soon

Link [here](https://github.com/moj-analytical-services/splink/compare/v3.9.2...v3.9.3#diff-2f4d3d45a1a46bfb35455d5665926f073e41a64f93e24a5e42b50dbb2661b44aL150), which doesn't seem to take you immediately to the correctly line, but it's here:

![image](https://github.com/moj-analytical-services/splink/assets/2608005/f70037e3-4520-48b5-acc0-301228bf37de)

**Comment by RobinL** [comment]
Deliving a bit deeper:
- There used to be a link type called [`"link_only_find_matches_to_new_records"`](https://github.com/moj-analytical-services/splink/blob/4a803c2952b9d27ad708522995ea85111f4cf9c6/splink/linker.py#L1793C42-L1793C79).  The `'link_only` part because it doesn't find links within the new records, just between them.
- The `find_matches_to_new_records` is a inner join as follows:

  ```
  select ...
  from __splink__df_concat_with_tf as l
  inner join __splink__df_new_records_with_tf as r
  on 1=1 where 1=1 
  ```
- Note that the link type is not respected - e.g. if `source_dataset` is populated in the new records, it's ignored for hte purpose of a `link_only` i.e. the user is expected to cut down the results 
- In the working 3.9.2 version, the user didn't have to provide a 'source_dataset', and this didn't appear in the results, for this reason.

**Comment by RobinL** [comment]
Worth noting the source dataset is strictly needed for the labelling tool because for a link_only it needs access to the source dataset column

It's also useful to have the source dataset column autopopulated because it makes it clearer what's an original record and what's a new record

**Comment by RobinL** [comment]
@mshearer0 to get working code you can add a 'source_dataset' to your `record`
```
record = {
'source_dataset': "new record",
'unique_id': 1,
 'Postcode': "BH15 4QE",
 'CompanyName': "VANTAGE YACHT RECRUITMENT",
}
```

But there's a proper fix incoming in #1416
